### ISSUE TITLE
Fixing some a11y, need React guidance

### ISSUE THREAD

**Issue by svinkle** [freeCodeCamp/guides#897]
Hi there! I'm attempting to fix up some accessibility issues with the site itself and need some help with understanding this React code. Specifically, the elements which render the side nav panels.

I've narrowed it down to [`renderHeader()`](https://github.com/freeCodeCamp/guides/blob/master/src/LayoutComponents/nav/NavPanel.jsx#L76) where this issue seems to be. This method returns a `div` with some `span` elements, but on component render, the `span`s are children of an `a` element. These `a` elements are using the appropriate `aria-*` attributes to help convey state and context, but there are a couple issues:

1. The `a` elements are missing an `href` in order to receive keyboard focus
2. The `onClick` handler is on the parent `div` element. This needs to be set on the `a` in order for keyboard users to activate the panel.

I can't figure out where or how the `a` elements are being rendered? Would this change even be possible?

I'm fixing up a few other things which I'll send along in a PR, but this one's got me in a loop. Any insight to where this change would be made would be appreciated!

**Comment by Ethan-Arrowood** [comment]
HI @svinkle could you inspect element on the exact elements you are referencing and post a screenshot here please? This will help me as I traverse the React structure.

**Comment by svinkle** [comment]
Ah yes, that would have beeb helpful.

Here you go, @Ethan-Arrowood:

![screen shot 2017-10-15 at 11 06 49 am](https://user-images.githubusercontent.com/1392632/31586001-0c046d80-b199-11e7-91aa-7344577f411f.png)

**Comment by Ethan-Arrowood** [comment]
The NavPanel is made up of individual React-Bootstrap [Panel](https://react-bootstrap.github.io/components.html#panels) elements. The <a> element is created by this component. I'd recommend investigating the [Panel](https://github.com/react-bootstrap/react-bootstrap/blob/master/src/Panel.js) source code. Try to find the parameter the <Panel /> component passes to the <a> href property. Once you figure that out, update the code in NavPanel.js where we are rendering the <Panel /> element. 

Give it a try, and if you can't find it I can help you more.

**Comment by svinkle** [comment]
Got it! Good pointers, @Ethan-Arrowood. Thanks!

**Comment by Ethan-Arrowood** [comment]
You should keep this issue open, edit the code, and submit a PR and reference this issue. Once your PR is merged we can close the issue. When you do submit a PR let me know I'd be happy to review it for you 👍

**Comment by svinkle** [comment]
Ah yes, good flow there. Will do.

**Comment by Ethan-Arrowood** [comment]
Hows this going?

**Comment by svinkle** [comment]
@Ethan-Arrowood Don't worry I haven't forgotten about you! I need to break up my work into smaller PRs then I'll send them over for review. I have time coming up this weekend/early next week so expect some things then!

**Comment by svinkle** [comment]
@Ethan-Arrowood there's one more thing I'd like to do but also need some guidance.

I'm trying to create a `ref` from a wrapper element in `templates/Article.jsx`. Then in `LayoutComponents/nav/NavItem.jsx`, on `Link` click, send focus to the `ref`.

By shifting keyboard focus this will help keyboard-only and screen reader users to get from the side nav to the article area quickly.

Any tips on creating the `ref` and passing it to the `NavItem` component?

**Comment by Ethan-Arrowood** [comment]
@svinkle where is this `<article />` tag? If it is in a parent component of `NavItem` then you can pass the ref as a prop down to the `NavItem`. If its a sibling I'd recommend lifting it up to the parent element and then passing it back down to the `NavItem`. This project also uses Redux so you may want to consider adding it to the store. @Bouncey may be able to share more insight on the best implementation of this attribute. Are these changes being tracked on some PR? If I can see the rest of the code it may be easier for me to give you pointers.

**Comment by svinkle** [comment]
@Ethan-Arrowood Right, the `article` element is something that I'm adding in `Article.jsx` No PR yet, just working on it locally in a new branch.

I can't figure out where or how `<Article/>` is being used as it's not being imported anywhere. If I knew this I might be able to figure out how to pass it around and back down to `NavItem`.

**Comment by Ethan-Arrowood** [comment]
Going to pin this to my todo list and will revisit it on Sunday when I have some time. Keep me posted if you make any progress before then.

**Comment by Bouncey** [comment]
@svinkle this is where `gatsby` comes in. 

Article is passed to the Layout component as props. We don't import Article directly, it is used as a template for the HTML coming from the markdown files

As for refs. You should be able to do everything in Layout, you will be able to update the ref in a lifecycle hook like componentDidUpdate or componentWillReceiveProps. 

If you are going to wrap `props.children` in an article tag, it should be able to handle none article routes, like `/search` and not use the article tag.

**Comment by svinkle** [comment]
@Bouncey The `article` element was originally a `div` so, it should already do what you're expecting.

I've tried to figure this `ref` part out, but no dice. My React knowledge is limited. I'm not sure how to do this as both `Article` and `Layout` components are not extending `React.Component` as a class; they're both just functions. Can you still use lifecycle hook methods here?

Any other tips or pointers would be really great.

**Comment by Bouncey** [comment]
Convert whichever you need into a class, I have no issues with this.

**Comment by sw-yx** [comment]
hey Scott - coming here from twitter. just want to offer some react guidance but i havent dived into this specific code. basically what i see is you are trying to pass info from a child comment back up to a parent. there are a few ways to do this: 1) use redux, 2) use parent methods that you pass down as props, 3) use refs. 

i think refs are a messy creature in react and should be avoided as far as possible, but if you're gonna use them anyway check out the docs <https://reactjs.org/docs/refs-and-the-dom.html> in particular you forgot a `this.` as a first order error

**Comment by AlmeroSteyn** [comment]
Hi @svinkle .

Also from your twitter question. Honestly I did not have the time to look at the actual solution (night time here now) but from what you say sounds like a pretty standard setting focus on route transition.

Maybe there are reasons in the app architecture for not being able to do this but would it be possible to convert the `article` component to a class component? If it is being rendered as a wrapper for the content  for each article it should be as simple as:

```
class MyArticle extends Component{

componentDidMount(){
    this.article.focus();
}

render(){
        return (
           <article ref={article => this.article = article}>
                    //Some content
            </article>
        )
}
}
```

The rules for refs are:
1. They can only be used with DOM elements and React classes. With DOM elements you get a reference to the DOM element itself and with React classes you get the class instance of the component. This, of course, means you can use refs to DOM elements in functional components.
2. The ref callback is executed on component mount time so then you should store the reference and you can act on it imperatively later, when required, as above.

Personally I would not be shy to convert from function to React class if the need arrises. There seems to be a swing in the community from the whole use classes for god components and functions for view. If you need it use it, is my opinion. It beats passing the thing down the component tree (or worse, putting it in Redux) when it can be an internal component matter.

I used the same thing here https://github.com/AlmeroSteyn/react-a11y-patterns/blob/master/src/components/PageFocusSection.js

If you do want to act on the ref in another component you can pass the callback above as a prop to the component:

```
const HasRef = ({articleRef}) => <article ref={articleRef>//Some content</article>;

class UsesRef extends Component {

componentDidMount(){
    this.article.focus();
}

render(){
        return (
           <HasRef articleRef={article => this.article = article}>
                    //Some content
            </HasRef>
        )
}
}
```

This way you pass the ref callback as a prop (as deep as you want) and you will get the element you want to focus anywhere in the tree.

I would NOT be afraid to use refs here. This is one of the very reason React provides this escape hatch. To imperatively focus elements. This is even confirmed in the docs itself regarding refs :-)

If this does not help I will be happy to see if I can apply it in the actual code example tomorrow, time permitting :-)

PS. In a Redux application, ALWAYS use the pattern above to pass the ref through as setting a ref on a class component wrapped with the react-redux connect HOC will return a reference to the HOC itself and not the wrapped component. The callback pattern above means that the prop is passed through to the wrapped component and it, in effect, bypasses the connect HOC and you get a handle to the internal DOM element you want to focus, as expected.

**Comment by AlmeroSteyn** [comment]
@svinkle I had a look at the actual codebase this morning.

From what I can see there does not seem to be a way, aside from the Redux implementation, to pass het the ref of `article` to the `NavItem` component as they are siblings of the same component tree.

The following implementation of `article` will focus the `article` element on each navigation from the menu:
```
class Article extends Component {
    componentDidMount() {
      this.article.focus();
    }

    render() {
      const article = this.props.data.markdownRemark;
      const { pathname } = this.props.location;
      const {
        html,
        fields: {
          slug
        },
        frontmatter: {
          title
        }
      } = article;
      return (
        <div>
          <Helmet>
            <title>{ `${title} | freeCodeCamp Guide` }</title>
            <link
              href={ `https://guide.freecodecamp.org${slug}` }
              rel='canonical'
            />
          </Helmet>
          <Breadcrumbs path={ pathname } />
          <article
            dangerouslySetInnerHTML={{ __html: html }}
            ref={article => {this.article = article;}}
            tabIndex='-1'
          />
        </div>
      );
    }
}
```

The only drawback I could see was that when you refresh the page, or navigate to a deeplinked page, it will also focus itself. The question would be if this is a problem.

If so and the `ref` has to be focussed on navigation click you will probably need to store the current article `ref` in Redux and then focus it in the click handler of the link. Although, somehow that feels a bit weird :-)

I also thought about using hash links to send focus to the `article` element. Matching up the `id` of the `article` element and the hash portion of the link URL should do the focussing automatically. Tried it out but see there are already hashes being used and don't have the time now to investigate further to come up with a neat generic solution in this app.

Hope this helps. Would love to hear if it did, and if not, why as focus management in React is an interest of mine and always interested in more use cases in real world apps.

**Comment by svinkle** [comment]
@AlmeroSteyn, thanks for the deep dive! Will dig in tonight or tomorrow and let you know how it goes. 👍

**Comment by AlmeroSteyn** [comment]
@svinkle cool!

Actually in https://github.com/AlmeroSteyn/react-a11y-patterns/blob/master/src/components/PageFocusSection.js I got around the constant focus on refresh by sending state with React Router.  Forgot to mention that in the previous comment.

```
componentDidMount() {
    const { location } = this.props;
    if (location.state && location.state.setFocus) {
      this.header.focus();
    }
  }
```

This way you can control when focus is set. You could maybe do the same think if gatsby link passes through the state.

**Comment by svinkle** [comment]
@Bouncey @AlmeroSteyn I couldn't figure this out for the life of me, so I had to get a little sneaky about it, but it works. See what you think of #7818.

**Comment by AlmeroSteyn** [comment]
@svinkle 

Dove a little bit deeper and you are quite right. The Gatsby Link component seems to disallow the option to pass an object to `to` in the code. Normally, with React Router you can do this:

```
<Link to="{pathname: path, state: {flag: true}}">Demo link</Link>
```

But `Gatsby Link` passes the `to` prop into this function internally before setting it as the component state:

```
export function withPrefix(path) {
  return normalizePath(pathPrefix + path)
}
```
So it expects the `to` to only be a `string`.

This IS normally the way to send state from the link to the routed component with React Router, would not know immediately why that is disallowed in Gatsby.

One can probably solve this by registering the clicks in Redux, but that seems like quite a complex solution just to do this task.

Your solution, while very imperative, seems to be a very reasonable workaround seeing the limits of the Link component. I would just use it, unless Gatsby Link can be changed to accept state as well.

**Comment by Bouncey** [comment]
If this is the way to go. I'd be happy to accept a forked version of `gatsby-link` that accepts objects. You should maybe open a PR against the gatsbyjs/gatsby repo also 👍

**Comment by Ethan-Arrowood** [comment]
Impressive work @svinkle great job on this!

**Comment by svinkle** [comment]
Thanks, @Ethan-Arrowood!

@AlmeroSteyn, right; my primary issue was that I couldn't pass a prop of some state back down to `<Article/>` for the `componentDidMount()` conditional check.

That said, @Bouncey are your expecting a different solution from me based on @AlmeroSteyn's note, or is this 👍?

**Comment by Bouncey** [comment]
Which ever way you would like to tackle the issue, I am agnostic about it right now :+1: 

Thanks for sticking with this @svinkle

**Comment by svinkle** [comment]
Well, the way I see it:

1. It works as is, and
2. I need to put this to bed and move on.

When #7818 gets merged feel free to close out this ticket. It was great working with y'all! 🙂
